bookkeeper-server/src/main/java/org/apache/bookkeeper/proto/BookieClientImpl.java

```diff
@@ -407,14 +409,16 @@ static ChannelReadyForAddEntryCallback create(
         @Override
         public void operationComplete(final int rc,
                                       PerChannelBookieClient pcbc) {
-            if (rc != BKException.Code.OK) {
-                bookieClient.completeAdd(rc, ledgerId, entryId, addr, cb, ctx);
-            } else {
-                pcbc.addEntry(ledgerId, masterKey, entryId,
-                              toSend, cb, ctx, options, allowFastFail, writeFlags);
+            try {
+                if (rc != BKException.Code.OK) {
+                    bookieClient.completeAdd(rc, ledgerId, entryId, addr, cb, ctx);
+                } else {
+                    pcbc.addEntry(ledgerId, masterKey, entryId,
+                            toSend, cb, ctx, options, allowFastFail, writeFlags);
+                }
+            } finally {
+                ReferenceCountUtil.release(toSend);
             }
-
-            ReferenceCountUtil.release(toSend);
             recycle();
```

Comment:
should the `recycle()` also be in the `finally` block?

Reply:
I think it's better to not recycle objects if exceptions are thrown. Recycling is optional and omitting it won't cause any issues since exceptions are hopefully rare.